### ISSUE TITLE
Allow to unit test and detect an attempt to `fx.Populate` non-pointer types

### ISSUE THREAD

**Issue by alexykot** [uber-go/fx#762]
We're using `fx` for wireup of both long-living servers and short-living CLI commands, and using `fx.Populate` in the wireup of CLI commands to extract and use CLI dependencies. 

We're also using `fx.ValidateApp` to unit test the completeness of the wireup, and it works pretty well in most situations. It does not work for one specific problem though - when the targets provided to `fx.Populate` are unsuitable for populating (e.g. non-pointer objects).

Consider this piece of wireup:
```
func cliDepOptions(app *microservice.App, dep ...interface{}) fx.Option {
	return fx.Options(
		coreOptions(app),
		fx.Populate(dep),
	)
}
```
This produces a set of core options that are needed for any CLI command, and it populates the supplied list of dependencies. There may be different dependencies for different CLI commands, so we use variadic `...interface{}` for it. 

This wireup is subtly wrong, and this cannot be detected by compiler or the unit tests using `fx.ValidateApp`. `fx.Populate(dep)` should really be `fx.Populate(dep...)`. Both will compile, but the first version effectively supplies a single `[]interface{}` object, and that one cannot be populated.

This situation cannot be unit tested because of the interplay between how the `fx.Populate` is designed and how `dig.DryRun` option works. `fx.Populate()` returns an `fx.Option`, and in case of a failure this option is an invoker that takes no arguments and simply returns the detected error, it looks like this - `Invoke(func() error { return err })`. 

However, in a unit test, with `dig.DryRun` option enabled, no invokers are triggered so the `fx.New` does not receive the error produced by `fx.Populate` and this mistake goes unnoticed.

As a result - humans writing the CLI wireup can make a small mistake that will go unnoticed by both the compiler and the unit tests, and will only be caught later in the integration testing or after deployment. 

I'm not sure what would be the best approach here, whether it's an `fx` or `dig` issue, but would be good to fix this and allow these situations to be unit tested.

**Comment by alexykot** [comment]
Seems like the solution for this would be not that complicated. 

In case of an error the `Populate()` call should return an `errorOption` instead of `invokeOption`. That will allow the app to return an error even if `dig` container is in `dryRun` mode and all dependencies are met.

I would be happy to make a PR for this if this approach seems reasonable.

**Comment by abhinav** [comment]
Resolved in #784. Thanks for the fix, @alexykot!
